lib/spack/spack/build_environment.py

```diff
@@ -813,158 +801,242 @@ def setup_package(pkg, dirty, context="build"):
+    a flag specifying in what way they do so. The list is ordered topologically
+    from root to leaf, meaning that environment modifications should be applied
+    in reverse so that dependents override dependencies, not the other way around."""
+    visitor = traverse.TopoVisitor(
+        EnvironmentVisitor(*specs, context=context),
+        key=lambda x: x.dag_hash(),
+        root=True,
+        all_edges=True,
+    )
+    traverse.traverse_depth_first_with_visitor(traverse.with_artificial_edges(specs), visitor)
+
+    # Dictionary with "no mode" as default value, so it's easy to write modes[x] |= flag.
+    modes = defaultdict(lambda: 0)
+    nodes_with_type = []
+
+    for edge in visitor.edges:
+        key = edge.spec
+
+        # Mark the starting point
+        if edge.parent is None:
+            modes[key] = ROOT
+            continue
+
+        mode = modes[edge.parent]
+
+        # Nothing to propagate.
+        if not mode:
+            continue
+
+        deptypes = edge.depflag
+
+        # Dependending on the context, include particular deps from the root.
+        if ROOT & mode:
+            if context == Context.BUILD:
+                if (dt.BUILD | dt.TEST) & deptypes:
+                    modes[key] |= BUILDTIME_DIRECT
+                if dt.LINK & deptypes:
+                    modes[key] |= BUILDTIME
+
+            elif context == Context.TEST:
+                if (dt.RUN | dt.TEST) & deptypes:
+                    modes[key] |= RUNTIME_EXECUTABLE
+                elif dt.LINK & deptypes:
+                    modes[key] |= RUNTIME
+
+            elif context == Context.RUN:
+                if dt.RUN & deptypes:
+                    modes[key] |= RUNTIME_EXECUTABLE
+                elif dt.LINK & deptypes:
+                    modes[key] |= RUNTIME
+
+        # Propagate RUNTIME and RUNTIME_EXECUTABLE through link and run deps.
+        if (RUNTIME | RUNTIME_EXECUTABLE | BUILDTIME_DIRECT) & mode:
+            if dt.LINK & deptypes:
+                modes[key] |= RUNTIME
+            if dt.RUN & deptypes:
+                modes[key] |= RUNTIME_EXECUTABLE
+
+        # Propagate BUILDTIME through link deps.
+        if BUILDTIME & mode:
+            if dt.LINK & deptypes:
+                modes[key] |= BUILDTIME
+
+        # Finalize the spec; the invariant is that all in-edges are processed
+        # before out-edges, meaning that edge.parent is done.
+        if not (ADDED & mode):
+            modes[edge.parent] |= ADDED
+            nodes_with_type.append((edge.parent, mode))
+
+    # Attach the leaf nodes, since we only added nodes
+    # with out-edges. Can this be improved?
+    for spec, mode in modes.items():
+        if mode and not (ADDED & mode):
+            nodes_with_type.append((spec, mode))
+
+    return nodes_with_type
+
+
+class SetupContext:
+    """This class encapsulates the logic to determine environment modifications, and is used as
+    well to set magic globals in modules of package.py."""
```

Comment:
(minor) Let's not talk about :magic_wand: :
```suggestion
    well to set global names in modules of package.py."""
```